test/rule_creation_protocol/protocol.py

```diff
@@ -0,0 +1,181 @@
+
+    @classmethod
+    def get_parsed_value_fields(cls, parsed_name) -> dict:
+        return {
+            field_name: parsed_name[field_name]
+            for field_name, field_info in cls.model_fields.items()
+            if getattr(field_info, 'json_schema_extra') and getattr(field_info, 'json_schema_extra').get('parsed_value', False)
+        }
+
+    @model_validator(mode='before')
+    def parse_name(cls, values) -> dict:
+        name = values.get('name')
+        if name is None:
+            raise ValueError('Name is required')
+
+        parser = ParsePattern()
+        parsed_name = parser.parse_feature_name(name)
+        if not parsed_name:
+            raise ProtocolError(
+                value=name,
+                message="Please check: the convention name for the name is: /n\
+                    - A functional part (such as 'ALB') /n\
+                    - A separator (e.g. a '-') /n\
+                    - The rule title (e.g. 'Alignment in spatial structure /n\
+                    - In between each single words, there must a separator"
+            )
+        values.update(cls.get_parsed_value_fields(parsed_name))
+        return values
+
+    @field_validator('rule_code')
+    def validate_rule_code(cls, value) -> dict:
+        """ The rule code is made of 3 digits capital letters (taken from the list of Functional parts)
+        + 3 digit numbers 
+        e.g. 'ALB123', 'GEM391' and not 'AB14' or 'GKE9001'
+        """
+        validator_helper = ValidatorHelper()
+        validation = validator_helper.validate_rule_code(value)
+        if validation:
+            raise ProtocolError(
+                value=validation['value'],
+                message=validation['message']
+            )
+        return value
+
+    @field_validator('first_separator')
+    def validate_first_separator(cls, value, values):
+        valid_first_separator = values.data['valid_first_separator']
+        if value != valid_first_separator:
+            raise ProtocolError(
+                value=value,
+                message=f"expected separator {valid_first_separator} but instead found {value}"
+            )
+        return value, values
+    
+    @field_validator('separators')
+    def validate_separators(cls, value, values):
+        separators = value
+        valid_separators = values.data['valid_separators']
+        unvalid_separators = [separator for separator in separators if separator != valid_separators]
+        if any(unvalid_separators):
+            raise ProtocolError(
+                value = separators,
+                message = f"expected {valid_separators} but found the following unvalid seperators {unvalid_separators}"
+            )
+        return value, values
+
+
+class RuleCreationConventions(ConfiguredBaseModel):
+    """Validate feature conventions
+    """
+    feature: Naming  # feature as in .feature file
+    dotfeature_file: Naming  # either user input or test file
```

Comment:
As far as I noticed, those are the only spots where the `Naming` class is created. I thought of the grammar `feature: Naming` as type hinting only and was quite surprised to see it's the way the class is invoked. 

Is it a pydantic pattern? Or am I missing something?

Reply:
Both :)

I've created a seperate Naming pydantic class because a lot of 'Naming' is equal in terms of implementation
GEM001 - Closed shell edge usage 
GEM001_Closed-shell-edge-usage.feature

What is tested here are the seperators; `-` `_` and ` `, `-`, respectively. I thought it would be easier to create a seperate pydantic class for this. In which the following elements are validated:

- Before : This is used to parse the name in elements that are tested seperately. If something goes wrong with parsing, an error is raised. The parsing is done with pyparsing with a pre-defined structure (First the rule code, then a seperator, than a collection of words and an optional .feature). The parsed results are then tested by the other `field_validators` 
- Rule code : test whether 'GEM001' is correct. And not GE001 or GEM01
- First seperators and second seperators. Checks if `_` is used between GEM001 and Closed-shell-edge usage and not `-` or `!`